### DISCUSSION TITLE
Investigate need for shared code base for Java desktop, webapp and mobile apps

### DISCUSSION THREAD

**Discussion by chimp1984** [bisq-network/bisq2#351] [General]
Related to https://github.com/bisq-network/bisq2/discussions/298 and https://github.com/bisq-network/bisq2/discussions/350 we should investigate if we need a shared code base for supporting several application targets.

### Targets
- Java desktop application (full featured reference implementation) for all major dekstop OS.
- Semi-full node Android version. New Android studio project with Kotlin UI using bisq2 Java libraries as dependencies. The chat based trade protocol is the primary target. Not clear if other trade protocols make sense due to more complex dependencies. Does not run the gossip network but use a relay node to connect to the P2P network (can be a user controlled or provided relay node). I2P is not strictly required to be supported.
- Web application for self-hosted node projects (Umbrel, BTC-pay server,...). Requires to run a Java headless full node which provides a REST api. Web application is a thin UI application using the REST api for all core functionality.
- Light mobile app for iOS. As Tor is problematic to integrate on iOS and I2P cannot be supported as its Java, we limit the iOS mobile app to a remote control for a full nodes (either the users desktop UI app or a headless app).  

### Technology choices
For the Android version it is Kotlin (or Java) for the UI. The core functionality is using the Java libraries from the bisq2 repo.

The Web application and the light iOS mobile app will be mainly UI and a minimal application structure as well as the Rest api client. I think the best results are achieved with native code: Html/JS for web and Swift for iOS. Code re-use does not really make much sense as all core funtionality is provided by the REST api and comes from the Java full node.

There might be one option where code re-use might make sense. A semi-full node for iOS using https connections to a relay node instead of Tor. The private keys for the identies can be hosted on the mobile app, so the relay cannot not decrypt any messages. It only relays public messages (e.g. offers) and encrypted private messages (trade messages, mailbox messages,...). It learns about the IP of the client and number of messages but not more.

With long polling we could emulate push data. The app would request all pending data at activation.
For supporting such a model we would need to re-use the Java code base. Kotlin Multiplatform requires Kotlin when targeting iOS. It would be quite some effort to port all to Kotlin but it would be do-able. We also would need a facade for the network API to use either the full P2P network code base or the https based client-server based approach.

My main concern with targeting iOS is that we are depending on Apples closed garden policy and live on their mercy to permit such applications and in light of increased pressure on privacy apps (EU trying to block E2E encryption) I would not be too surprised if Apple would kick out such apps and then all the invested effort would be wasted. Apple has acted like that back in 2013 by kicking out all Bitcoin apps, so that risk is real. In that light I am skeptical to invest into such an effort for iOS. The light iOS app seems much less of effort (mainly porting Android UI to Swift which should not be too hard as Kotlin and Swift are very similar).
We should also keep in mind that in emerging markets most people have Android and those who have Apple phones are usually wealthy enough to have a laptop or desktop computer.

We can keep that decision also for later and see how politics develops and how successful the mobile app on Android is, to see if the risk of investing in a semi-full iOS app is acceptable. The required efforts of adding a facade to the network module and porting all code to Kotlin would be significant already now and we don't have the dev resources to do that. So I would suggest to drop that option for now and re-evaluate it when the time is right for it.

**Comment by salokiam** [comment]
regarding Web application for self-hosted node projects: That Installation will be full fledge anyway, so running a Servlet container should not be an issue, since this is a server enviornment anyway. Instead of html/JS we could use a all Java Web Framework to write the web application. That way the JavaFX programmers have a easier time to catch up / read the web-app code and we some add so many technologies to our project. See also my other [comment ](https://github.com/bisq-network/bisq2/discussions/350#discussioncomment-2877231)

**Comment by xineohp2021** [comment]
@chimp1984 
Thanks for directing me to this thread. As discussed, I will be picking up the project of developing the base Android application using a Android Studio (and a mix of Android and Kotlin). If there is a breakthrough which allows iOS development, I will let you know (and you can also notify me of same).

I will create a proposal under [Bisq 1 Proposals](https://github.com/bisq-network/proposals/issues) as there appears not to be an existing one for Bisq 2; or will you create a separate Proposal/Issue page for  Bisq 2?

And as discussed, I will be putting up progress reports on that proposal thread at the end of every week.

Thank you and I look forward to working with you.

Regards.
